src/exodus/sigmawallet.cpp

```diff
@@ -0,0 +1,597 @@
+    }
+
+    // HMAC-SHA512(SHA256(count), key)
+    std::array<unsigned char, CSHA256::OUTPUT_SIZE> countHash;
+
+    std::vector<unsigned char> result;
+    result.resize(CSHA512::OUTPUT_SIZE);
+
+    auto count = std::to_string(GetSeedIndex(seedId));
+    CSHA256().
+        Write(
+            reinterpret_cast<const unsigned char*>(count.data()),
+            count.size()
+        ).
+        Finalize(countHash.data());
+
+    CHMAC_SHA512(countHash.data(), sizeof(countHash)).
+        Write(key.begin(), key.size()).
+        Finalize(result.data());
+
+    seed = uint512(result);
+
+    return GetSeedIndex(seedId);
+}
+
+bool GetChildFromPath(string const &path, std::pair<uint32_t, bool> &child)
+{
+    auto startPos = path.find_last_of('/') + 1;
+    if (startPos >= path.size()) {
+        return false;
+    }
+
+    auto childData = path.substr(startPos);
+
+    auto isHardened = childData.back() == '\'';
+    if (isHardened) {
+        childData.resize(childData.size() - 1);
+    }
+
+    if (childData.empty()) {
+        return false;
+    }
+
+    auto childPos = std::stol(childData);
+
+    if (childPos > std::numeric_limits<uint32_t>::max()) {
+        return false;
+    }
+
+    child = {static_cast<uint32_t>(childPos), isHardened};
+
+    return true;
+}
+
+uint32_t SigmaWallet::GetSeedIndex(CKeyID const &seedId)
+{
+    if (!pwalletMain->mapKeyMetadata.count(seedId)) {
+        throw std::runtime_error("key not found");
+    }
+
+    auto const &meta = pwalletMain->mapKeyMetadata[seedId];
+
+    // parse last index
+    std::pair<uint32_t, bool> child;
+    if (!GetChildFromPath(meta.hdKeypath, child)) {
+        throw std::runtime_error("fail to parse HD key path");
+    }
+
+    if (child.second) {
+        throw std::runtime_error("hardened is not allowed");
+    }
+
+    return child.first;
+}
+
+bool SigmaWallet::SeedToPrivateKey(
+    const uint512& seedZerocoin, exodus::SigmaPrivateKey& coin)
+{
+    //convert state seed into a seed for the private key
+    uint256 nSeedPrivKey = seedZerocoin.trim256();
+    nSeedPrivKey = Hash(nSeedPrivKey.begin(), nSeedPrivKey.end());
+
+    // generate serial and randomness
+    sigma::PrivateCoin priv(sigma::Params::get_default(), sigma::CoinDenomination::SIGMA_DENOM_1);
+    priv.setEcdsaSeckey(nSeedPrivKey);
+
+    // Create a key pair
+    secp256k1_pubkey pubkey;
+    if (!secp256k1_ec_pubkey_create(OpenSSLContext::get_context(), &pubkey, priv.getEcdsaSeckey())) {
+        return false;
+    }
+
+    // Hash the public key in the group to obtain a serial number
+    auto serialNumber = priv.serialNumberFromSerializedPublicKey(OpenSSLContext::get_context(), &pubkey);
+
+    //hash randomness seed with Bottom 256 bits of seedZerocoin
+    Scalar randomness;
+    auto randomnessSeed = ArithToUint512(UintToArith512(seedZerocoin) >> 256).trim256();
+    randomness.memberFromSeed(randomnessSeed.begin());
```

Comment:
Just `randomness.memberFromSeed(seedZerocoin.begin() + 32)` is enough. It is also easy to understand how bytes was manipulated too.

Reply:
this logic is also work with other situation which is running on big-endian machine and `>>` with a number which is not multiple of 32. Changing to `randomness.memberFromSeed(seedZerocoin.begin() + 32)` eliminate that functionality then I disagree to change.

Reply:
After discussion the safest way is to treat it as a byte array so we can go with my idea.